File: Haskell/data_structures/09-RabinKarp.hs
import Data.Char (ord)
import Data.Vector (Vector(..), fromList, generate, (!))
import qualified Data.Vector as V
import System.Random (randomR, mkStdGen)


polynomialHashing :: Vector Char -> Integer -> Integer -> Int
polynomialHashing s prime x = fromInteger $ foldr (\chr acc -> (acc * x + toInteger (ord chr)) `mod` prime) 0 s


calcPowerOfX :: Integer -> Integer -> Int -> Integer
calcPowerOfX x prime exponent = foldl (\acc _ -> (acc * x) `mod` prime) 1 [1..exponent]


substringHashes :: Vector Char -> Int -> Int -> Integer -> Integer -> Vector Int
substringHashes text textLength patternLength prime x =
    let len = textLength - patternLength + 1
        xPower = calcPowerOfX x prime patternLength
        hashValues = generate len calcHashes
        calcHashes :: Int -> Int
        calcHashes ix =
            if ix == len - 1 then
                let lastSubstring = V.slice (len - 1) patternLength text in polynomialHashing lastSubstring prime x
            else
                let hashed = x * toInteger (hashValues ! (ix + 1))
                               + toInteger (ord (text ! ix)) - xPower * toInteger (ord (text ! (ix + patternLength)))
                in fromInteger $ hashed `mod` prime
    in hashValues


getMatchingIndices :: Vector Char -> Int -> Vector Char -> Int -> Int -> Vector Int -> [Int]
getMatchingIndices text textLength pattern patternLength patternHash hashValues =
    let len = textLength - patternLength + 1
        go :: Int -> [Int] -> [Int]
        go ix acc
            | ix == len = reverse acc
            | patternHash == hashValues ! ix && pattern == V.slice ix patternLength text = go (ix + 1) (ix : acc)
            | otherwise = go (ix + 1) acc
    in go 0 []


findPatternInText :: Vector Char -> Vector Char -> Integer -> [Int]
findPatternInText text pattern prime =
    let x = fst $ randomR (1, prime - 1) (mkStdGen 2112)
        patternLength = V.length pattern
        textLength = V.length text
        patternHash = polynomialHashing pattern prime x
        hashValues = substringHashes text textLength patternLength prime x
    in getMatchingIndices text textLength pattern patternLength patternHash hashValues


main :: IO()
main = do
    pattern <- fmap fromList getLine
    text <- fmap fromList getLine
    let prime = 1000000007
    let indices = findPatternInText text pattern prime
    putStrLn $ unwords $ map show indices

File: Julia/data_structures/09-Rabin-Karp.jl
using Random


polynomialhash(s, prime, x) = foldr((char, acc) -> (acc*x + Int(char)) % prime, s; init=0)

powerofX(x, prime, k) = foldl((acc, _) -> acc*x % prime, 1:k; init=1)


function precompute_substring_hashes(text, textlength, patternlength, p, x)
    hashes = zeros(Int, textlength-patternlength+1)
    lastsubstring = text[end-patternlength+1:end]
    hashes[end] = polynomialhash(lastsubstring, p, x)
    xpower = powerofX(x, p, patternlength)
    for ix in (textlength-patternlength):-1:1
        hashes[ix] = (x*hashes[ix+1] + Int(text[ix]) - xpower*Int(text[ix+patternlength])) % p
    end
    hashes
end


function rabinkarp(text, pattern, prime)
    x = Random.rand(1:(prime-1))
    patternlength = length(pattern)
    textlength = length(text)
    matched = Int[]
    h = polynomialhash(pattern, prime, x)
    substringhashes = precompute_substring_hashes(text, textlength, patternlength, prime, x)
    for ix = 1:(textlength-patternlength+1)
        if h == substringhashes[ix] && pattern == text[ix:ix+patternlength-1]
            push!(matched, ix - 1)
        end
    end
    matched
end


function main()
    pattern = readline() |> collect
    text = readline() |> collect
    prime = 1000000007
    Random.seed!(2112)
    result = rabinkarp(text, pattern, prime)
    println(join(result, " "))
end


main()
